sksports/model/heart_rate.py

```diff
@@ -0,0 +1,228 @@
+
+    hr_pred = np.ones(power.size) * hr_start
+    for curr_idx in range(hr_pred.size):
+        hr_prev = hr_pred[curr_idx - 1] if curr_idx > 0 else hr_start
+        hr_pred[curr_idx] = (hr_prev +
+                             (hr_ss[curr_idx] - hr_prev) / rate[curr_idx])
+
+    return hr_pred
+
+
+def _model_residual(params, heart_rate, power):
+    """Residuals between prediction and ground-truth."""
+    return heart_rate - exp_heart_rate_model(power, *params)
+
+
+def _model_residual_least_square(params, heart_rate, power):
+    """Least-square error of the residuals."""
+    return np.sum(_model_residual(params, heart_rate, power) ** 2)
+
+
+class HeartRateRegressor(BaseEstimator, RegressorMixin):
+    """Estimate the heart-rate from power data.
+
+    The model used is based on the formulation in [1]_.
+
+    Parameters
+    ----------
+    hr_start : float, default=75
+        Initial heart-rate frequency.
+
+    hr_max : float, default=200
+        Heart-rate frequency of the athlete at maximum.
+
+    hr_slope : float, default=0.30
+        Slope considered if the model is linear. FIXME
+
+    hr_drift : float, default=3e-5
+        Attenuation of the heart-rate other time. FIXME
+
+    rate_growth : float, default=24
+        Growth rate of the exponential when the power increases.
+
+    rate_decay : float, default=30
+        Decay rate of the exponential when the power decreases.
+
+    Attributes
+    ----------
+    hr_start_ : float
+        Fitted initial heart-rate frequency.
+
+    hr_max_ : float
+        Fitted heart-rate frequency of the athlete at maximum.
+
+    hr_slope_ : float
+        Fitted slope considered if the model is linear. FIXME
+
+    hr_drift_ : float
+        Fitted attenuation of the heart-rate other time. FIXME
+
+    rate_growth_ : float
+        Fitted growth rate of the exponential when the power increases.
+
+    rate_decay_ : float
+        Fitted decay rate of the exponential when the power decreases.
+
+    References
+    ----------
+    .. [1] de Smet, Dimitri, et al. "Heart rate modelling as a potential
+       physical fitness assessment for runners and cyclists." Workshop at ECML
+       & PKDD. 2016.
+
+    """
+
+    def __init__(self, hr_start=75, hr_max=200, hr_slope=0.30, hr_drift=3e-5,
+                 rate_growth=24, rate_decay=30):
+        self.hr_start = hr_start
+        self.hr_max = hr_max
+        self.hr_slope = hr_slope
+        self.hr_drift = hr_drift
+        self.rate_growth = rate_growth
+        self.rate_decay = rate_decay
+
+    def _check_inputs(self, X, y=None):
+        """Validate the inputs.
+
+        Check if the inputs are Pandas inputs. Resample with a sampling-rate of
+        1 Hertz.
+
+        """
+        if y is None:
+            is_df = True if hasattr(X, 'loc') else False
+            X_ = X.resample('1s').mean() if is_df else X
+            X_ = check_array(X_)
+            return X_.ravel(), is_df
```

Comment:
You still need to return `y` here as the second return argument right? Else line 193 will raise a `ValueError` (unpack error).

Reply:
Since it is confusing, I should always return `X_.ravel(), y, is_df` and ignore `y` when it is None.
Right now, I am returning different thing depending if it is `fit` or `predict`.

Reply:
Ah now I get it. That could work indeed but I think it's less confusing if a method always returns the same number of arguments.